### ISSUE TITLE
SA1008 (OpeningParenthesisMustBeSpacedCorrectly) incorrectly reported for cast in foreach

### ISSUE THREAD

**Issue by sharwell** [DotNetAnalyzers/StyleCopAnalyzers#683]
The following code incorrectly results in a SA1008 warning, claiming that no space should exist before the `(` character of the cast.

```csharp
foreach (object element in (IEnumerable)collection)
{
}
```

**Comment by AArnott** [comment]
So does this:

    Run(async () =>
    {
    }|);

**Comment by AArnott** [comment]
And this:

    SomeMethod(delegate () { });

**Comment by sharwell** [comment]
Blocked on #193.

**Comment by vweijsters** [comment]
> And this:
> 
> ```C#
> SomeMethod(delegate () { });
> ```

This should actually report a diagnostic. StyleCop reports this as well, and I think its correct. The delegate keyword here should be seen as a method declaration and those don't allow a space between the method name and the parameter list either.
